### DISCUSSION TITLE
LDM Notes for May 1st, 2023

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#7167] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2023/LDM-2023-05-01.md

## Agenda

- Fixed Size Buffers
- `lock` statement improvements

**Comment by TahirAhmadov** [comment]
Re. `lock`, I'm confused - so are any changes going to be made to the language at this point or not?

**Reply by CyrusNajmabadi** [reply]
> Conclusion

> The language will support using the new locking pattern for locking on System.Lock.

**Reply by TahirAhmadov** [reply]
You won't believe this, but I'm still confused :) Is the `lock` going to be expanded to "auto-switch" to use `Lock` type at compile time?

**Reply by CyrusNajmabadi** [reply]
No. But you'll be able to say `lock(_lockOfTheNewType)` and have that be new semantics (not Monitor semantics).

**Reply by TahirAhmadov** [reply]
OK that's the same thing I meant. It's disappointing that we're going with an incremental change, one that has issues (albeit not terrible ones), before even having a plan. Literally it says "experiment" and yet the language is being changed for it. 
Oh wel...

**Reply by CyrusNajmabadi** [reply]
I'm not sure what you mean.  All we're saying is: if the runtime comes up with a new, preferred, way of locking with a type in the BCL, then we will support it language-wise.

**Reply by jnm2** [reply]
Tahir is referring to the notes here:
> This new type, on the other hand, is part of an experiment to entirely replace monitor-based locks in .NET, and the hope is that new development will use these instead of an object to perform mutual exclusion.

**Reply by TahirAhmadov** [reply]
I meant, I think it would be better to not make any changes to the language, specifically, not to change `lock` to infer the type and change from `Monitor` to "surface area", before a comprehensive plan can be devised.

**Reply by CyrusNajmabadi** [reply]
I'm not sure how that's any different than what we're doing.

**Reply by CyrusNajmabadi** [reply]
We're saying: we will align the language with the plans the runtime makes here as long as that continues to make sense.

**Reply by Joe4evr** [reply]
@CyrusNajmabadi I'm guessing that includes: "*If* the runtime decides to push the plan out (for whatever reason), the language won't change here"?

**Reply by CyrusNajmabadi** [reply]
Yes. If this is clearly a feature that depends on that happening. If it doesn't happen... What would the feature even do?

**Reply by TahirAhmadov** [reply]
So, if the `Lock` type isn't added (for whatever reason), then there will be no language changes; if it is added, the `lock` will start to recognize it.
That's the thing - I think it would be better to wait for real world experience of using that new type using `using` (ha ha ha), before making any changes to the language. Like others suggested, an analyzer can be made available together with the new type to warn about its usage in `lock` so that users aren't mistakenly expecting it to work that way, for the time being.
When (if) a decision is confidently made that `Monitor` can be retired over time, language changes can be discussed then, IMO.

**Reply by jnm2** [reply]
There might be a misunderstanding. Not every decision in LDM has to necessarily end up in a future unified C# spec just because a decision was made. For example, `!!`. In one sense, anything could be thought of as an experiment till it ships in GA with no preview flag.

**Reply by TahirAhmadov** [reply]
@jnm2 That's fine, my thought process here is that C# shouldn't even begin experimenting before the runtime has a plan for the entire vertical. 
If the plan is to slowly convert `lock` to use the new "surface area" lock types, that's fine, I can live that (in fact, I can see benefits in that approach). The problem is we don't know what the plan is; it looks like there is no plan, other than "we'll make minor language changes in response to the runtime team's introduction of new types".

**Reply by jnm2** [reply]
The C# design team's input is potentially quite valuable to the the runtime team as part of their own design process. Can you explain the issue you have in mind with this?

**Reply by TahirAhmadov** [reply]
My concern is with this sentence:
> An open question is whether we want to convince people to avoid monitor-based locks entirely; that seems like it would be best driven by the runtime, rather than the language design team.

That gives me absolutely no confidence whether I should start using the new locking types, what's going to happen in the future with `Monitor` the type, lockability of all objects, etc. (even if `Lock` is added to .NET vNext and `lock` uses it in C# vNext).

On the other hand, if locking type(s) were added, specifically aimed at `using`, I'd be perfectly happy using those in a few places to look at how they perform, as part of an ecosystem-wide experiment.

My confidence level aside, what if the surface area is not sufficient for whatever reason? Now we're "locked into" `lock` (haha) language construct, and changing it becomes more difficult.

**Reply by CyrusNajmabadi** [reply]
> That gives me absolutely no confidence 

This is all an experiment currently. I'm not sure what you're looking for :-)

Our position simply is that we feel like if the runtime modernizes locks that we would bring the language along with that, that's all.

**Reply by HaloFour** [reply]
> Our position simply is that we feel like if the runtime modernizes locks that we would bring the language along with that, that's all.

I think that's a reasonable position.  Monitor-based locking as a feature baked into every reference always felt like a mistake copied from Java to me.  If the intent is to explore moving towards a more intentional design that feels like a good plan.  However, my concern is that Monitor locks are _so_ entrenched in the runtime and ecosystem that this will be impractical and that we'll be left in some halfway state.

**Comment by TonyValenti** [comment]
Is there I link I can read about the lock experiment in more detail?

Also, I would like it if lock had a short form like using does. 

lock objlock;

**Reply by 333fred** [reply]
https://github.com/dotnet/runtime/issues/34812

We did briefly mention the short form, but afaik there's no championed proposal for it.

**Comment by orthoxerox** [comment]
I remember how confused I was by nullable jagged arrays when I had to consciously realize that the innermost brackets mean the outermost array. But I am confused again.

> This is then complicated further by nullability: a non-null regular array of 4 element int nullable arrays would be `int[]?[4]`

Isn't that a nullable regular array of 4-element non-null arrays of non-null ints?

**Reply by TahirAhmadov** [reply]
Nope, that would be a 4-element inline array (which cannot be nullable by definition) of nullable regular arrays of non-null ints. 
Or maybe not? I'm not sure now :)

**Reply by colejohnson66** [reply]
`int[]?[]` currently means, an array (of unspecified size) of `int[]?`. Basically, read jagged arrays backwards. This is evident by the initialization syntax: `new int[]?[a]` means an array of length `a` of `int[]?` (`int[]` or `null`).

As such, `int[]?[4]` would mean an array with a fixed size of 4 of `int[]?` (instead of an unspecified size).

IMO, the simplest solution is to avoid jagged arrays directly if at all possible. If you need a jagged array, see if the inner array (the `int[]?` here) could be put into a class/struct. Then you'd have `MyContainer?[]`, and it becomes significantly easier to reason about. If you're storing a table, use a 2D array instead (`int[,] arr = new int[a,b];`).

**Reply by 333fred** [reply]
> As such, int[]?[4] would mean an array with a fixed size of 4 of int[]? (instead of an unspecified size)

Nope. As I called out in the notes, this is not what this means. It's an array of unspecified size of nullable `int[4]`s.

**Reply by colejohnson66** [reply]
But that's backwards from how jagged arrays work, so why propose it like that?

**Reply by HaloFour** [reply]
> It's an array of unspecified size of nullable `int[4]`s.

aka `Nullable<Buffer4<int>>[]`?

**Reply by orthoxerox** [reply]
This is why Go moved the brackets to the front of the type.

And now I think I am going crazy, because that's what the compiler accepts right now and that's now what I remember:

```c#
int?[][] arr1 = new[] { new [] { (int?)null }}; //array of array of nullable int
int[]?[] arr2 = new[] { (int[]?)null }; //array of nullable array of int
int[][]? arr3 = (int[][]?)null; //nullable array of array of int
```

So it turns out, in the type definition, we read the brackets from the outside in. In the dereferencing, we read them from the inside out:

```c#
string[][] arr2x1 = new[] {new[] {"(1,1)"}, new[] {"(2,1)"}};
Console.WriteLine(arr2x1[1][0]);

string[]?[] arr2x1n = new[] {(string[]?)null, new[] {"(2,1)"}};
Console.WriteLine(arr2x1n[1]?[0]);
```

So if I wanted to change this into a fixed-sized definition, it looks to me that I would have to write

```c#
string[1][2] arr2x1 = new[] {new[] {"(1,1)"}, new[] {"(2,1)"}};
Console.WriteLine(arr2x1[1][0]);

string[1]?[2] arr2x1n = new[] {(string[]?)null, new[] {"(2,1)"}};
Console.WriteLine(arr2x1n[1]?[0]);
```

**Reply by HaloFour** [reply]
@orthoxerox 

> So if I wanted to change this into a fixed-sized definition, it looks to me that I would have to write

No, I think that would be `string[2][1]`, or `Buffer2<Buffer1<string>>`

**Reply by orthoxerox** [reply]
Gah!

```c#
string[]?[] arr2x1n = new string[2][];
arr2x1n[0] = null;
arr2x1n[1] = new string[] { "2,1" };
```

So, it's nullability from right to left and lengths from left to right. It's not just a footgun, it's a GAU-8 footautocannon.

**Reply by CyrusNajmabadi** [reply]
But has also been this way for several years now, and tends to be an annoyance but not much more.  So it's one of those C#1.0 warts you live with and think about fixing in some new language :)

**Reply by orthoxerox** [reply]
Yeah, it's an annoyance when one is the type of the variable and another is the type of the value. But mixing them in fixed-size buffer declarations has already blown several minds of C# enthusiasts and not just regular C# users.

**Reply by CyrusNajmabadi** [reply]
> has already blown several minds of C# enthusiasts 

I mean... the same was true when we did nullable originally.  My point is that it didn't turn out to be a substantial problem in practice.  It's just a wart, not a foot gun.

**Reply by HaloFour** [reply]
>But has also been this way for several years now

Longer than that if you consider that NVTs initially set the tone back in C# 2.0 with `int?[][]`.  NRTs didn't really have anywhere else to go.

**Reply by 333fred** [reply]
> Gah!
> 
> ```cs
> string[]?[] arr2x1n = new string[2][];
> arr2x1n[0] = null;
> arr2x1n[1] = new string[] { "2,1" };
> ```
> 
> So, it's nullability from right to left and lengths from left to right. It's not just a footgun, it's a GAU-8 footautocannon.

Yup. @colejohnson66, this is what I mean. C# very unfortunately copied C here, and the syntax does not work how you think it does. And, imo, it's one thing if you can only observe this inconsistency in a `new`, it's quite another if you can observe it in any type expression.

**Reply by colejohnson66** [reply]
In that case, maybe const generics should become a thing. Afterwards, instead of shoehorning fixed size buffers into array syntax, we'd use a `Buffer<byte, 4>`? Feels clunky, but it'd be more understandable how nesting works.

**Reply by Joe4evr** [reply]
@colejohnson66 You're going to make @Sergio0694 happy with that. 🤣
